Compute items for SqrtDetSpatialMetric

prayush opened the pull request:
## Proposed changes
**Edit**: after merging #1468, this PR has been updated to:
 - add a compute tag that returns the `sqrt` of the determinant of 3-metric;
 - add a generic `Sqrt` compute tag that can be used on `Scalar<>`.

**Previous description** (before merging #1468):
Adds compute tags for Det of SpatialMetric, SqrtDet of SpatialMetric, and InverseSpatialMetric. This PR includes changes added in #1468 as the first commit. Therefore, please review the **most recent** commit only.

### Types of changes:

- [ ] Bugfix
- [x] New feature

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests, `clang-tidy` and `IWYU`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

prayush committed:
Add compute item for sqrtdet3metric, and a generic Sqrt compute item

carmaza commented on src/PointwiseFunctions/GeneralRelativity/ComputeSpacetimeQuantities.hpp:
```diff
@@ -256,5 +256,68 @@ struct DetAndInverseSpatialMetricCompute
+ *
+ * \details Can be retrieved using `gr::Tags::SqrtDetSpatialMetric`.
+ */
+template <size_t SpatialDim, typename Frame, typename DataType>
```
I've also encountered compute times that are just the square root of other items. We should probably create a generic square root compute item.

prayush replied:
@carmaza : I added one via 1c99b0eb1bae71e19d8701df02675bcfc2096d35 . (see its unit test for use case) Was this what you meant?

carmaza replied:
Yes, this is what I was thinking of. I'll let the @sxs-collaboration/spectre-core-devs check if they're happy with the location of this feature. Also, I'm not sure about how to deal with the tag names of "Sqrt" tags ending in "Squared" (Sqrt_LapseSquared, etc...). Any thoughts?

prayush replied:
I would think that we have any `MyTag` whose square-root we will tag as `Sqrt<MyTag>` and thats it. We need to either tag the sqrt as `Sqrt,>` or we add a `Squared` or similar postfix, but I don't think both. But, maybe I am misunderstanding your question?

carmaza replied:
Nevermind. I was thinking of quantities whose square is easier to compute first (like the sound speed squared in hydro), but that might be just too specific to consider.

carmaza commented on src/PointwiseFunctions/GeneralRelativity/ComputeSpacetimeQuantities.hpp:
```diff
@@ -256,5 +256,68 @@ struct DetAndInverseSpatialMetricCompute
+
+/*!
+ * \ingroup GeneralRelativityGroup
+ * \brief Compute item for inverse spatial metric \f$\g^{ij}\f$
```
g^{ij}

carmaza commented on src/PointwiseFunctions/GeneralRelativity/Tags.hpp:
```diff
@@ -38,6 +38,11 @@ struct InverseSpatialMetric : db::SimpleTag {
   static std::string name() noexcept { return "InverseSpatialMetric"; }
 };
 template <typename DataType>
+struct DetSpatialMetric : db::SimpleTag {
```
Test the tag name.

prayush replied:
see [this similar comment](https://github.com/sxs-collaboration/spectre/pull/1468#discussion_r277442773)

carmaza commented on tests/Unit/PointwiseFunctions/GeneralRelativity/Test_ComputeSpacetimeQuantities.cpp:
```diff
@@ -157,6 +157,15 @@ SPECTRE_TEST_CASE("Unit.PointwiseFunctions.GeneralRelativity.SpacetimeDecomp",
+                                              DataVector>::name() ==
+        "SqrtDetSpatialMetric");
+  CHECK(
+      gr::Tags::InverseSpatialMetric<3, Frame::Inertial, DataVector>::name() ==
```
`gr::Tags::InverseSpatialMetricCompute`

carmaza reviewed (commented):
You can squash the fixups. Also, I'm not sure if Github is being weird with the commit order, but the new commit should go before in the history so each commit compiles independently.

kidder commented on src/DataStructures/Tensor/EagerMath/Magnitude.hpp:
```diff
@@ -106,4 +105,27 @@ struct Normalized : db::ComputeTag {
 };
+
+/// \ingroup TensorGroup
+/// \brief Compute the Euclidean magnitude of a rank-0 tensor
```
huh?  the magnitude of a scalar is just its absolute value...

kidder commented on src/PointwiseFunctions/GeneralRelativity/ComputeSpacetimeQuantities.hpp:
```diff
@@ -256,5 +256,68 @@ struct DetAndInverseSpatialMetricCompute
+
+/*!
+ * \ingroup GeneralRelativityGroup
+ * \brief Compute item to get the square root of the determinant of the spatial
```
This just gets the determinant...

nilsdeppe commented:
IWYU failed.

kidder commented:
I am confused, this only has `SqrtDetSpatialMetric` and `Sqrt`;  are there commits missing or is the commit message wrong?

prayush commented:
> I am confused, this only has `SqrtDetSpatialMetric` and `Sqrt`; are there commits missing or is the commit message wrong?

@kidder - TL;DR: the commit message was obsoleted, I've updated it. 

... what happened was that in #1468, originally there was a compute item being added that returned `std::pair<DetSpatialMetric, InverseSpatialMetric`, and this PR had 2 compute items that extracted those 2 variables out from inside `std::pair`. However, #1468 was changed to return a `Variables<DetSpatialMetric, InverseSpatialMetric>` instead, which made those 2 compute items in this PR obsolete. I deleted those two, and now this PR only adds compute tag for getting `Sqrt` of `DetSpatialMetric`.

nilsdeppe commented on src/DataStructures/Tensor/EagerMath/Magnitude.hpp:
```diff
@@ -106,4 +105,28 @@ struct Normalized : db::ComputeTag {
+/// \details
+/// Computes the square root of the absolute value of the scalar.
+template <typename DataType>
+Scalar<DataType> sqrt_magnitude_of_scalar(
```
- Why is this function in the `Tags` namespace?
- why `_of_scalar`? That's given just by the argument type, so I don't think that's necessary.

prayush replied:
agreed, fixed.

nilsdeppe commented on tests/Unit/PointwiseFunctions/GeneralRelativity/Test_ComputeSpacetimeQuantities.cpp:
```diff
@@ -165,14 +168,18 @@ SPECTRE_TEST_CASE("Unit.PointwiseFunctions.GeneralRelativity.SpacetimeDecomp",
-  for (size_t i = 1; i <= 3; ++i) {
-    expected_spacetime_metric.get(i, i) += 1.;
-  }
+  const auto expected_spacetime_metric = [&]() {
```
you should only capture what you need, `&generator, &distribution, &used_for_size`

prayush replied:
done

nilsdeppe commented:
@prayush this also has conflicts now. Feel free to squash everything when you rebase :)

nilsdeppe reviewed (approved):
LGTM! :+1:
